File: it/src/main/scala/org/zalando/nakadi/client/scala/test/factory/EventIntegrationHelper.scala
package org.zalando.nakadi.client.scala.test.factory

import scala.concurrent.duration.DurationInt

import org.scalatest.Matchers
import org.scalatest.WordSpec
import org.zalando.nakadi.client.scala.Client
import org.zalando.nakadi.client.scala.Client
import org.zalando.nakadi.client.scala.ClientError
import org.zalando.nakadi.client.scala.model._
import org.zalando.nakadi.client.scala.model._
import org.zalando.nakadi.client.scala.model._
import org.zalando.nakadi.client.scala.model.Event
import org.slf4j.LoggerFactory
import scala.concurrent.Await
import scala.concurrent.duration.DurationInt
import scala.concurrent.Future

class EventIntegrationHelper(generator: EventGenerator, client: Client) extends WordSpec with Matchers {
  private val log = LoggerFactory.getLogger(this.getClass)
  val eventType = generator.eventType

  def createEventType(): Boolean = {
    wasItsuccessfull(executeCall(client.createEventType(eventType)), "createEventType")
  }

  def getEventType(eventTypeName: String = eventType.name): Option[EventType] = executeCall(client.getEventType(eventTypeName)) match {
    case Left(clientError) =>
      wasItsuccessfull(Option(clientError), "getEventType")
      None
    case Right(result) => result
  }

  def deleteEventType() = wasItsuccessfull(executeCall(client.deleteEventType(eventType.name)), "deleteEventType")

  def publishEvents(nrOfEvents: Int): Seq[Event] = {
    val events = for {
      a <- 1 to nrOfEvents
    } yield generator.newEvent()
    wasItsuccessfull(executeCall(client.publishEvents(eventType.name, events)), "publishEvents")
    log.info(s"EVENTS published: $events")
    events
  }

  def updateEventType(eType: EventType): Boolean = {
    wasItsuccessfull(executeCall(client.updateEventType(eventType.name, eType)), "updateEventType")
  }

  def eventTypeExist(eventTypeName: String = eventType.name): Boolean = {
    getEventType(eventTypeName) match {
      case None    => false
      case Some(_) => true
    }
  }

  def getNumberOfPartitions(): Int = extractFromRightOptional(() => client.getPartitions(eventType.name)).size

  private def extractFromRightOptional[T](function: () => Future[Either[ClientError, Option[T]]]) = {
    val received = executeCall(function())
    assert(received.isRight == true, s"ClientErrror ${received}")
    val Right(opt) = received
    assert(opt.isDefined == true, "because it expected Some, but received None")
    val Some(eventTypes) = opt
    eventTypes
  }


  def getEnrichmentStrategies(): Seq[EventEnrichmentStrategy.Value] = extractFromRightOptional(() => client.getEnrichmentStrategies())
  def getPartitionStrategies(): Seq[PartitionStrategy.Value] = extractFromRightOptional(() => client.getPartitioningStrategies())

  //Private methods
  private def wasItsuccessfull(in: Option[ClientError], msg: String): Boolean = in match {
    case Some(clientError) =>
      log.error(s"${generator.eventTypeId} - ${msg} => ClientError: $clientError")
      false
    case None =>
      true
  }

  private def executeCall[T](call: => Future[T]): T = {
    Await.result(call, 10.second)
  }

}

 

File: it/src/main/java/org/zalando/nakadi/client/java/test/event/generator/EventIntegrationHelper.java
package org.zalando.nakadi.client.java.test.event.generator;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.ExecutionException;
import java.util.stream.IntStream;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.zalando.nakadi.client.java.Client;
import org.zalando.nakadi.client.java.enumerator.PartitionStrategy;
import org.zalando.nakadi.client.java.model.Event;
import org.zalando.nakadi.client.java.model.EventType;
import org.zalando.nakadi.client.java.model.Partition;

public class EventIntegrationHelper {

    private final EventGenerator gen;
    private final Client client;
    private final EventType eventType;
    private Logger log = LoggerFactory.getLogger(this.getClass());

    public EventIntegrationHelper(EventGenerator gen, Client client) {
        this.gen = gen;
        this.client = client;
        this.eventType = gen.getEventType();
    }

    public Optional<EventType> getEventType() {
        try {
            return client.getEventType(gen.getEventTypeName()).get();
        } catch (InterruptedException | ExecutionException e) {
            log.error(gen.getEventTypeId() + " - createEventType :" + e.getMessage(), e);
            return Optional.empty();
        }
    }

    public boolean createEventType() {
        try {
            client.createEventType(eventType).get();
            return true;
        } catch (RuntimeException | InterruptedException | ExecutionException e) {
            logError(gen.getEventTypeId() + " - createEventType :" + e.getMessage(), e);
            return false;
        }
    }

    @SuppressWarnings("static-access")
    private void logError(String identifier, Throwable t) {
        log.error("%s => %s ".format(identifier, t.getMessage()));
    }

    public EventGenerator getGen() {
        return gen;
    }

    public Client getClient() {
        return client;
    }

    public Integer getNumberOfPartitions() throws InterruptedException, ExecutionException {
        Optional<List<Partition>> result = client.getPartitions(eventType.getName()).get();
        return result.get().size();
    }
    public List<PartitionStrategy> getPartitionStrategies() throws InterruptedException, ExecutionException {
        return client.getPartitioningStrategies().get().get();
        
    }

    public List<Event> publishEvents(Integer nrOfEvents) throws InterruptedException, ExecutionException {
        List<Event> events = new ArrayList<Event>();
        IntStream.range(0, nrOfEvents).forEach(nr -> events.add(gen.getNewEvent()));
        client.publishEvents(eventType.getName(), events).get();
        log.info("Events published" + events.size());
        return events;

    }

}
